Fix C++ compile warnings

```
 89%] Built target page_table_test
In file included from /Users/lei/work/lance/cpp/src/lance/format/schema_test.cc:2:
/Users/lei/work/lance/cpp/src/lance/arrow/testing.h:44:42: warning: unused parameter 'storage_type' [-Wunused-parameter]
      std::shared_ptr<::arrow::DataType> storage_type, const std::string& serialized) const override {
                                         ^
/Users/lei/work/lance/cpp/src/lance/arrow/testing.h:77:42: warning: unused parameter 'storage_type' [-Wunused-parameter]
      std::shared_ptr<::arrow::DataType> storage_type, const std::string& serialized) const override {
                                         ^
In file included from /Users/lei/work/lance/cpp/src/lance/arrow/scanner_test.cc:29:
/Users/lei/work/lance/cpp/src/lance/arrow/testing.h:44:42: warning: unused parameter 'storage_type' [-Wunused-parameter]
      std::shared_ptr<::arrow::DataType> storage_type, const std::string& serialized) const override {
                                         ^
/Users/lei/work/lance/cpp/src/lance/arrow/testing.h:77:42: warning: unused parameter 'storage_type' [-Wunused-parameter]
      std::shared_ptr<::arrow::DataType> storage_type, const std::string& serialized) const override {
                                         ^
[ 90%] Linking CXX executable type_test
[ 91%] Linking CXX executable metadata_test
[ 91%] Built target type_test
[ 91%] Built target metadata_test
[ 91%] Linking CXX executable plain_test
2 warnings generated.
[ 92%] Linking CXX executable schema_test
In file included from /Users/lei/work/lance/cpp/src/lance/arrow/writer_test.cc:31:
/Users/lei/work/lance/cpp/src/lance/arrow/testing.h:44:42: warning: unused parameter 'storage_type' [-Wunused-parameter]
      std::shared_ptr<::arrow::DataType> storage_type, const std::string& serialized) const override {
                                         ^
/Users/lei/work/lance/cpp/src/lance/arrow/testing.h:77:42: warning: unused parameter 'storage_type' [-Wunused-parameter]
      std::shared_ptr<::arrow::DataType> storage_type, const std::string& serialized) const override {
                                         ^
/Users/lei/work/lance/cpp/src/lance/arrow/scanner_test.cc:126:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  ::arrow::RegisterExtensionType(ext_type);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~
[ 92%] Built target plain_test
[ 92%] Built target schema_test
[ 93%] Linking CXX executable utils_test
/Users/lei/work/lance/cpp/src/lance/arrow/writer_test.cc:266:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  arrow::RegisterExtensionType(ext_type);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~
```

Should we enforce `-Werror` for debug build?

--- a/cpp/CMakeLists.txt
+++ b/cpp/CMakeLists.txt
@@ -35,11 +35,12 @@ if (CCACHE_FOUND)
     set_property(GLOBAL PROPERTY RULE_LAUNCH_LINK ccache)
 endif (CCACHE_FOUND)
 
+add_compile_options(-Wall -Wextra -fPIC)
+
 if (CMAKE_BUILD_TYPE STREQUAL Debug)
     SET(LANCE_BUILD_BENCHMARKS ON)
+    add_compile_options(-Werror)
 endif ()
-
-add_compile_options(-Wall -Wextra -fPIC)
 
 # TODO: Fix on linux
 Include(FetchContent)

--- a/cpp/src/lance/arrow/scanner_test.cc
+++ b/cpp/src/lance/arrow/scanner_test.cc
@@ -123,7 +123,7 @@ std::shared_ptr<::arrow::dataset::Scanner> MakeScanner(std::shared_ptr<::arrow::
 TEST_CASE("Scanner with extension") {
   auto table = MakeTable();
   auto ext_type = std::make_shared<::lance::testing::ParametricType>(1);
-  ::arrow::RegisterExtensionType(ext_type);
+  CHECK(::arrow::RegisterExtensionType(ext_type).ok());
   auto scanner = MakeScanner(table);
 
   auto dataset = std::make_shared<::arrow::dataset::InMemoryDataset>(table);

--- a/cpp/src/lance/arrow/testing.h
+++ b/cpp/src/lance/arrow/testing.h
@@ -11,12 +11,12 @@
 //  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 //  See the License for the specific language governing permissions and
 //  limitations under the License.
-#include <arrow/type.h>
+
 #include <arrow/extension_type.h>
+#include <arrow/type.h>
 #include <fmt/format.h>
 
 #include <string>
-
 
 namespace lance {
 namespace testing {
@@ -41,12 +41,14 @@ class ImageType : public ::arrow::ExtensionType {
   }
 
   ::arrow::Result<std::shared_ptr<::arrow::DataType>> Deserialize(
-      std::shared_ptr<::arrow::DataType> storage_type, const std::string& serialized) const override {
+      [[maybe_unused]] std::shared_ptr<::arrow::DataType> storage_type,
+      const std::string& serialized) const override {
     if (serialized != "ext-struct-type-unique-code") {
       return ::arrow::Status::Invalid("Type identifier did not match");
     }
     return std::make_shared<ImageType>();
   }
+
   std::string Serialize() const override { return "image-ext"; }
 };
 
@@ -74,7 +76,8 @@ class ParametricType : public ::arrow::ExtensionType {
   }
 
   ::arrow::Result<std::shared_ptr<::arrow::DataType>> Deserialize(
-      std::shared_ptr<::arrow::DataType> storage_type, const std::string& serialized) const override {
+      [[maybe_unused]] std::shared_ptr<::arrow::DataType> storage_type,
+      const std::string& serialized) const override {
     const int32_t parameter = *reinterpret_cast<const int32_t*>(serialized.data());
     return std::make_shared<ParametricType>(parameter);
   }
@@ -89,5 +92,5 @@ class ParametricType : public ::arrow::ExtensionType {
   int32_t parameter_;
 };
 
-}
-}
+}  // namespace testing
+}  // namespace lance

--- a/cpp/src/lance/arrow/writer_test.cc
+++ b/cpp/src/lance/arrow/writer_test.cc
@@ -263,7 +263,7 @@ TEST_CASE("Write extension but read storage if not registered") {
 
 TEST_CASE("Extension type round-trip") {
   auto ext_type = std::make_shared<::lance::testing::ImageType>();
-  arrow::RegisterExtensionType(ext_type);
+  CHECK(arrow::RegisterExtensionType(ext_type).ok());
   auto table = MakeTable();
   auto sink = arrow::io::BufferOutputStream::Create().ValueOrDie();
   CHECK(lance::arrow::WriteTable(*table, sink).ok());
@@ -272,4 +272,3 @@ TEST_CASE("Extension type round-trip") {
   auto actual_table = ReadTable(sink);
   CHECK(table->Equals(*actual_table));
 }
-

